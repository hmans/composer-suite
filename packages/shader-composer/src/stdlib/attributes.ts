import { $ } from "../expressions"
import { GLSLType, Unit } from "../units"

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })
