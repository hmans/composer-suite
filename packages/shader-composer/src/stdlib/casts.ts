import { $ } from "../expressions"
import { type } from "../glslType"
import { GLSLType, Input, isUnit, Unit, UnitConfig } from "../units"
import { Float, Mat3, Mat4, Vec2, Vec3, Vec4 } from "./values"

export const float = (
  v: Input<"float" | "bool" | "int"> = 0,
  extras?: Partial<UnitConfig<"float">>
) => Float($`float(${v})`, extras)

export const vec2 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec2">>
) => Vec2($`vec2(${x}, ${y})`, extras)

export const vec3 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  z: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec3">>
) => Vec3($`vec3(${x}, ${y}, ${z})`, extras)

export const vec4 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  z: Input<"float"> = 0,
  w: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec4">>
) => Vec4($`vec4(${x}, ${y}, ${z}, ${w})`, extras)

/** Cast the given value to a mat3. */
export const mat3 = (i: Input<"mat3" | "mat4">) => Mat3($`mat3(${i})`)

/** Cast the given value to a mat4. */
export const mat4 = (i: Input<"mat3" | "mat4">) => Mat4($`mat4(${i})`)

export const unit = <T extends GLSLType>(i: Input<T>, config?: Partial<UnitConfig<T>>) =>
  isUnit(i) ? i : Unit(type(i), i, config)

/**
 * Wraps an input value into a unit that is configured to use a varying.
 *
 * @param i Input value (unit, JS value or expression) to wrap in a varying unit.
 * @param config Optional extra configuration for the newly created unit.
 * @returns A new unit that wraps the given value and is configured to use a varying.
 */
export const varying = <T extends GLSLType>(
  i: Input<T>,
  config?: Partial<UnitConfig<T>>
) => Unit(type(i), i, { ...config, varying: true })
