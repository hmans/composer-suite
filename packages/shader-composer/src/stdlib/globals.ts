import { $ } from "../expressions"
import { Int, Vec2 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const InstanceID = Int($`gl_InstanceID`, {
  name: "Instance ID",
  only: "vertex"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})
