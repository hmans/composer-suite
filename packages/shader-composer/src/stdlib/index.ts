/** @packageDocumentation
 * The `shader-composer/stdlib` module contains a core collection of Shader unit implementations.
 */

export * from "./artistic"
export * from "./casts"
export * from "./geometry"
export * from "./globals"
export * from "./logic"
export * from "./masters"
export * from "./math"
export * from "./rotation"
export * from "./scene"
export * from "./spaces"
export * from "./textures"
export * from "./uniforms"
export * from "./values"
export * from "./vectors"
