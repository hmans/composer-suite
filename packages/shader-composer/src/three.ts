export * from "@shader-composer/three"
