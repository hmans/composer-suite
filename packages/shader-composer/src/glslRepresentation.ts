import { Color, Matrix3, Matrix4, Vector2, Vector3, Vector4 } from "three"
import { isExpression } from "./expressions"
import { isSnippet } from "./snippets"
import { GLSLType, isUnit, Input, uniformName } from "./units"

export const glslRepresentation = (
  value: Input | undefined,
  typeHint?: GLSLType
): string => {
  if (value === undefined) return ""

  if (isUnit(value))
    return value._unitConfig.uniform ? uniformName(value) : value._unitConfig.variableName

  if (isExpression(value)) return value.render()

  if (isSnippet(value)) return value.name

  if (typeof value === "string") return value

  if (typeof value === "boolean") return value ? "true" : "false"

  if (typeof value === "number") {
    const s = value.toString()
    return typeHint === "int" ? s : s.match(/[.e]/) ? s : `${s}.0`
  }

  if (value instanceof Color) return `vec3(${g(value.r)}, ${g(value.g)}, ${g(value.b)})`

  if (value instanceof Vector2) return `vec2(${g(value.x)}, ${g(value.y)})`

  if (value instanceof Vector3) return `vec3(${g(value.x)}, ${g(value.y)}, ${g(value.z)})`

  if (value instanceof Vector4)
    return `vec4(${g(value.x)}, ${g(value.y)}, ${g(value.z)}, ${g(value.w)})`

  if (value instanceof Matrix3)
    return `mat3(${value
      .toArray()
      .map((n) => g(n))
      .join(", ")})`

  if (value instanceof Matrix4)
    return `mat4(${value
      .toArray()
      .map((n) => g(n))
      .join(", ")})`

  /* Fail */
  throw new Error(`Could not render value to GLSL: ${value}`)
}

const g = glslRepresentation
