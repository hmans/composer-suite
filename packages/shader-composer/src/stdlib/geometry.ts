/** @packageDocumentation
 * Geometryyyy
 */

import { Vector2 } from "three"
import { $, Expression } from "../expressions"
import { injectAPI, GLSLType, Input, Unit } from "../units"
import { localToViewSpace, localToWorldSpace } from "./spaces"
import { Bool, Mat4, Vec2, Vec3 } from "./values"

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3(localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3(localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions($`position`, "Vertex Position")
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const TilingUV = (
  uv: Input<"vec2"> = UV,
  tiling: Input<"vec2"> = new Vector2(1, 1),
  offset: Input<"vec2"> = new Vector2(0, 0)
) =>
  Vec2(
    $`vec2(
      ${uv}.x * ${tiling}.x + ${offset}.x,
      ${uv}.y * ${tiling}.y + ${offset}.y)`
  )

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })
