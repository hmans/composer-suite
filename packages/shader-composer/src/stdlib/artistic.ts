import { pipe } from "../pipes"
import { GLSLType, Input, Unit } from "../units"
import { VertexNormal, ViewDirection } from "./geometry"
import { Abs, Add, Lerp, Mul, Pow, Saturate, Smoothstep } from "./math"
import { Float } from "./values"
import { Dot } from "./vectors"

export type FresnelProps = {
  normal?: Input<"vec3">
  alpha?: Input<"float">
  bias?: Input<"float">
  intensity?: Input<"float">
  power?: Input<"float">
  factor?: Input<"float">
}

export const Fresnel = ({
  normal = VertexNormal.world,
  bias = 0,
  intensity = 1,
  power = 2,
  factor = 1
}: FresnelProps = {}) =>
  Float(
    pipe(
      normal,
      (v) => Dot(ViewDirection, v),
      (v) => Add(factor, v),
      (v) => Abs(v),
      (v) => Pow(v, power),
      (v) => Mul(v, intensity),
      (v) => Add(v, bias),
      (v) => Saturate(v)
    ),
    { name: "Fresnel" }
  )

export type GradientStops<T extends GLSLType = "vec3"> = GradientStop<T>[]

export type GradientStop<T extends GLSLType = "vec3"> = [Input<T>, Input<"float">]

export const Gradient = <T extends GLSLType = "vec3">(
  f: Input<"float">,
  ...stops: GradientStops<T>
) => {
  let color = stops[0][0]

  for (let i = 1; i < stops.length; i++) {
    const stop = stops[i]
    const previous = stops[i - 1]
    color = Lerp(color, stop[0], Smoothstep(previous[1], stop[1], f))
  }

  return color as Unit<T>
}
