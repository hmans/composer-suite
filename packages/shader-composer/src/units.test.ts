import { Vector2 } from "three"
import { compileShader } from "./compiler"
import { $ } from "./expressions"
import { glslRepresentation } from "./glslRepresentation"
import { Float, Vec2 } from "./stdlib"
import { Input, Unit } from "./units"

const glsl = glslRepresentation

describe("Unit", () => {
  it("creates a node of the specified type and value", () => {
    const v = Unit("float", 1)
    expect(v._unitConfig.type).toBe("float")
    expect(v._unitConfig.value).toBe(1)
    expect(glsl(v._unitConfig.value)).toBe("1.0")
  })

  it("supports string values (which will be used as verbatim expressions)", () => {
    const v = Unit("vec3", $`vec3(1.0, 1.0, 1.0)`)
    expect(glsl(v._unitConfig.value)).toBe("vec3(1.0, 1.0, 1.0)")
  })

  it("supports expression values", () => {
    const a = Float(1)
    const v = Unit("vec3", $`vec3(${a}, 1.0, 1.0)`)
    expect(glsl(v._unitConfig.value)).toBe(
      `vec3(${a._unitConfig.variableName}, 1.0, 1.0)`
    )
  })

  it("allows nodes to directly reference other nodes", () => {
    const source = Float(1)
    const v = Float(source)
    expect(v._unitConfig.value).toBe(source)
    expect(glsl(v._unitConfig.value)).toBe(source._unitConfig.variableName)
  })

  it("supports a 'varying' flag that will automatically make it pass its data as a varying", () => {
    const v = Unit(
      "float",
      $`1.0 + 2.0 + onlyAvailableInVertex.x`, // a value expression that can only work in a vertex shader
      {
        name: "A variable with a varying",
        varying: true,
        vertex: { body: $`value += 3.0;` } // Modiyfing the value before it gets put into the varying
      }
    )

    const [c] = compileShader(v)

    expect(c.vertexShader).toMatchSnapshot()
    expect(c.fragmentShader).toMatchSnapshot()
  })

  it("supports constructing nodes through constructor functions", () => {
    const Double = (f: Input<"float">) => Float($`(${f}) * 2.0`)
    const v = Double(1)
    expect(glsl(v._unitConfig.value)).toBe("(1.0) * 2.0")
  })

  it("constructor functions can pass string values as expressions", () => {
    const Double = (f: Input<"float">) => Float($`(${f}) * 2.0`)
    const v = Double($`5.0`)
    expect(glsl(v._unitConfig.value)).toBe(`(5.0) * 2.0`)
  })

  it("constructor functions can pass references to other nodes", () => {
    const Double = (f: Input<"float">) => Float($`(${f}) * 2.0`)
    const a = Float(1)
    const v = Double(a)
    expect(glsl(v._unitConfig.value)).toBe(`(${a._unitConfig.variableName}) * 2.0`)
  })

  it("constructor functions can pass expression values to other nodes", () => {
    const Double = (f: Input<"float">) => Float($`(${f}) * 2.0`)
    const a = Float(5)
    const v = Double($`${a} + 5.0`)
    expect(glsl(v._unitConfig.value)).toBe(`(${a._unitConfig.variableName} + 5.0) * 2.0`)
  })
})

describe("Vec2", () => {
  it("accepts a THREE.Vector2 as value", () => {
    const vec2 = Vec2(new Vector2(1, 2))
    expect(glsl(vec2._unitConfig.value)).toBe("vec2(1.0, 2.0)")
  })
})
