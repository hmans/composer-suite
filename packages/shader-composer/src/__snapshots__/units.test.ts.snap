// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit supports a 'varying' flag that will automatically make it pass its data as a varying 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

/*** UNIT: A variable with a varying ***/
varying float v_A_variable_with_a_varying_1;


void main()
{
  /*** UNIT: A variable with a varying ***/
  float A_variable_with_a_varying_1 = 1.0 + 2.0 + onlyAvailableInVertex.x;
  {
    float value = A_variable_with_a_varying_1;
    value += 3.0;
    A_variable_with_a_varying_1 = value;
  }
  v_A_variable_with_a_varying_1 = A_variable_with_a_varying_1;
  
  
}"
`;

exports[`Unit supports a 'varying' flag that will automatically make it pass its data as a varying 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

/*** UNIT: A variable with a varying ***/
varying float v_A_variable_with_a_varying_1;


void main()
{
  /*** UNIT: A variable with a varying ***/
  float A_variable_with_a_varying_1 = v_A_variable_with_a_varying_1;
  
  
}"
`;
